src/kvstore/HBaseEngine.cpp

```diff
@@ -0,0 +1,353 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "base/Base.h"
+#include "kvstore/HBaseEngine.h"
+#include "storage/KeyUtils.h"
+#include "dataman/NebulaCodecImpl.h"
+#include "network/NetworkUtils.h"
+#include "kvstore/KVStore.h"
+
+namespace nebula {
+namespace kvstore {
+
+using nebula::storage::KeyUtils;
+
+const char* kHBaseTableNamePrefix = "Nebula_Graph_Space_";
+const char* kSchemaVersionColumnName = "__sv__";
+const PartitionID kFillPartId = 0;
+const uint8_t kFillMin = 0x00;
+const uint8_t kFillMax = 0xFF;
+const size_t kMaxRowKeyLength = sizeof(PartitionID) + sizeof(VertexID)
+                              + sizeof(EdgeType) + sizeof(VertexID)
+                              + sizeof(EdgeRanking) + sizeof(EdgeVersion);
+
+HBaseEngine::HBaseEngine(GraphSpaceID spaceId,
+                         const HostAddr& hbaseServer,
+                         meta::SchemaManager* schemaMan)
+        : KVEngine(spaceId)
+        , schemaMan_(schemaMan) {
+    CHECK_NOTNULL(schemaMan_);
+    LOG(INFO) << "Connect to the HBase thrift server "
+              << network::NetworkUtils::intToIPv4(hbaseServer.first)
+              << ":" << hbaseServer.second;
+    client_ = std::make_unique<HBaseClient>(hbaseServer);
+    CHECK_NOTNULL(client_.get());
+    tableName_ = kHBaseTableNamePrefix + folly::to<std::string>(spaceId_);
+    partsNum_ = this->allParts().size();
+    // TODO(zhangguoqing) check/create the hbase table named tableName_.
+    // Now, the version 1.4.9 of HBase thrift2 does not support APIs for table operations.
```

Comment:
why `HBase 1.4.9` ? The latest one is `2.1.4`

Reply:
In my opinion, if we really want to use this feature in a real production environment, we should be most compatible with the widely used version rather than the latest version 2.1.4. As far as I know, hbase version 1.4.9 is widely used, and this version of thrift2 is also forward compatible with rel/1.4.0, and which is not substantially different from rel/2.1.4 [1].  The APIs for table operation is support after the branch-2.2 which is not released now.

```
[1] diff hbase.thrift149 hbase.thrift214
78c78, 80
< DELETE_COLUMNS = 1
---
> DELETE_COLUMNS = 1,
> DELETE_FAMILY = 2,
> DELETE_FAMILY_VERSION = 3
210a213,218
> enum TReadType {
> DEFAULT = 1,
> STREAM = 2,
> PREAD = 3
> }
>
229c237,238
< 14: optional bool small
---
> 14: optional TReadType readType
> 15: optional i32 limit
```

Reply:
Understood. According to your statement, we should create table from shell before using it .

Reply:
Yes, for unit tests, these operations[1] will be done by test runner.

[1] https://github.com/vesoft-inc/nebula/pull/413/files#diff-9d431e623cca5bcbd11ca24b91c18bb2R13